oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MultithreadedTraverseWithSortStrategy.java

```diff
@@ -408,6 +454,174 @@ public void run() {
+                    log.info("merge complete for {}", mergedFile.getName());
+                    return mergedFile;
+                }
+                log.error("merge failed for {}", mergedFile.getName());
+            } finally {
+                mergeTaskPhaser.arriveAndDeregister();
+            }
+
+            return mergedFile;
+        }
+    }
+
+    /**
+     * Class responsible for -
+     * <ol>
+     *     <li>Watching {@link #sortedFiles} for new sorted files</li>
+     *     <li>Submitting those files in batch to an {@link ExecutorService}</li>
+     *     <li>Collecting the results (sorted files) created by those tasks</li>
+     *     <li>Merge the result with any left over files to create a single sorted file</li>
+     * </ol>
+     * Strategy -
+     * <ol>
+     *      <li>Wait for n files</li>
+     *      <li>construct new list of files to be merged by checking if its already merged</li>
+     *    and create intermediate merge file
+     *    (if final merge) merge all intermediate merge files and create sorted file
+     *      <li>add all merged files to merged list</li>
+     * </ol>
+     */
+    private class MergeRunner implements Runnable {
+        private final ArrayList<File> mergedFiles = new ArrayList<File>();
+        private final ArrayList<File> unmergedFiles = new ArrayList<File>();
+        private final File sortedFile;
+        private final ExecutorService executorService;
+        private final int threadPoolSize = Integer.getInteger(PROP_MERGE_THREAD_POOL_SIZE, DEFAULT_NUMBER_OF_MERGE_TASK_THREADS);
+        private final int batchMergeSize = Integer.getInteger(PROP_MERGE_TASK_BATCH_SIZE, DEFAULT_NUMBER_OF_FILES_PER_MERGE_TASK);
+        private final Comparator fileSizeComparator = new SizeFileComparator();
+
+        public MergeRunner(File sortedFile) {
+            this.sortedFile = sortedFile;
+            this.executorService = Executors.newFixedThreadPool(threadPoolSize);
+        }
+
+        private List<File> getSmallestUnmergedFiles(int size) {
+            ArrayList<File> result = new ArrayList<File>(unmergedFiles);
+            result.remove(MERGE_POISON_PILL);
+            result.sort(fileSizeComparator);
+            int endIdx = size > result.size() ? result.size() : size;
+            return result.subList(0, endIdx);
+        }
+
+        private void markAsMerged(List<File> target) {
+            mergedFiles.addAll(target);
+            unmergedFiles.removeAll(target);
+        }
+
+        @Override
+        public void run() {
+            Phaser mergeTaskPhaser = new Phaser(1);
+            List<Future<File>> results = new ArrayList<>();
+            List<File> mergeTarget = new ArrayList<>();
+            int count = 0;
+
+            while (true) {
+                try {
+                    unmergedFiles.add(sortedFiles.take());
+                    if (sortedFiles.contains(MERGE_POISON_PILL) || unmergedFiles.contains(MERGE_POISON_PILL)) {
```

Comment:
Since this is a blocking queue now and we pick element by element, I don't see a benefit of doing 'contains` check. You could just check the element removed from the queue and break. That would simply this code and you would also not need to drain sortedFiles to unmergedFiles later.

Reply:
The reason I added that wide range check is so that it proceed directly to the final merge and we don't create anymore tasks if poison pill shows up.
Otherwise if there are 2000 files left to be merged, it would create more tasks and wait for those to finish before we can go for a final merge step which MIGHT(?) introduce more wait time.

Reply:
> Otherwise if there are 2000 files left to be merged, it would create more tasks and wait for those to finish

I think, actually it would be better to merge at most 64 files at a time (always the smallest 64). I think it's better than to merge 2000 files at the same time: that would result in a lower cache hit rate.